### PR TITLE
Fix build steps being stuck 'running'

### PR DESCRIPTION
**Pull request by smira**
As waiting for events was started after modifying a pod, it had a chance
to actually miss the event that step got finished.

### CODE REVIEW

**Comment by bradrydzewski** [comment]
I feel like this could still be subject to a race condition.  The goroutine could be schedule to execute after the start function is invoked. There is no guarantee it would execute in time

**Comment by smira** [comment]
yeah, that's true, but much better than it was before, I can work on a more robust version

File: engine/engine_impl.go [hunk 1]
```diff
@@ -205,51 +205,64 @@ func (k *Kubernetes) Run(ctx context.Context, specv runtime.Spec, stepv runtime.
 	return k.waitForTerminated(ctx, spec, step)
 }
 
-func (k *Kubernetes) waitFor(ctx context.Context, spec *Spec, conditionFunc func(e watch.Event) (bool, error)) error {
+func (k *Kubernetes) waitFor(ctx context.Context, spec *Spec, conditionFunc func(pod *v1.Pod) (bool, error)) error {
 	label := fmt.Sprintf("io.drone.name=%s", spec.PodSpec.Name)
 	lw := &cache.ListWatch{
 		ListFunc: func(options metav1.ListOptions) (k8sruntime.Object, error) {
```
Comment (bradrydzewski) [inline engine/engine_impl.go:211]: I'm not too familiar with the underlying code, but just wanted to double-check and make sure it is thread safe to modify the `metav1.ListOptions`?

**Comment by bradrydzewski** [comment]
thanks for the help, much appreciated!  I was wondering if we might want to put these changes behind a feature flag (temporarily) so that we can test it out in production for a while and make sure it handles all edge cases?  cc @sethpollack 

in terms of how to put behind a feature flag, I usually create different versions of the function (one with the new changes) and then invoke if the flag is set:

```text
if os.Getenv("DRONE_FEATURE_FLAG_SOME_NAME") == "true" {
  waitForNew(...)
} else {
  waitForOld(...)
}
```

File: engine/engine_impl.go (hunk 1 above)
Comment (smira) [reply engine/engine_impl.go:211]: I picked it up from the code in `kubectl`, I believe `Watch` process might set up some options we want to preserve when setting out label selector.

**Comment by smira** [comment]
@bradrydzewski I'm still playing with this runner, and unfortunately I still see builds stuck, from the logs it seems that we don't get event from K8s API about container being terminated (still shows container as running). 

So what I want to do really is to rewrite this watcher into single watch per pipeline. We have pipeline with 28 steps and high fan-out, so creating 28*2 watch loops doesn't sound like good use of resources when only one loop is enough.

File: engine/engine_impl.go [hunk 2]
```diff
@@ -205,51 +205,64 @@ func (k *Kubernetes) Run(ctx context.Context, specv runtime.Spec, stepv runtime.
...
-	return err
-}
+		pod, ok := obj.(*v1.Pod)
+		if !ok {
+			return true, fmt.Errorf("unexpected object type: %v", obj)
+		}
 
-func (k *Kubernetes) waitForReady(ctx context.Context, spec *Spec, step *Step) error {
-	return k.waitFor(ctx, spec, func(e watch.Event) (bool, error) {
+		return conditionFunc(pod)
```
Comment (sethpollack) [inline engine/engine_impl.go:237]: Why are we calling the `conditionFunc` in the `preconditionFunc`?

**Comment by sethpollack** [comment]
@smira I'm seeing a similar issue. The issue isn't that you aren't getting the event, if you describe the pod you will see that k8s still thinks the container is running.

File: engine/engine_impl.go (hunk 2 above)
Comment (smira) [reply engine/engine_impl.go:237]: that was my attempt to skip `Watch` altogether if `List` already returns the pod in the state we expect

**Comment by bradrydzewski** [comment]
@sethpollack so are you seeing the container fails to restart when we switch from the placeholder image to the actual image?  If I understand correctly, does this mean we would expect the pipeline step to fail with a kubernetes error message, as opposed to be stuck in a running state?

**Comment by sethpollack** [comment]
@bradrydzewski Yes, the container fails to restart when we switch from the placeholder image to the actual image. 

But no we would not expect a kubernetes failure because as far as kubernetes is concerned the container is running. If you describe the pod it has the image change but the container never restarted and is still in running state. If you look at the node the container is not there.

I think eventually you will get a kubernetes error, but the build times out before that happens.

**Comment by bradrydzewski** [comment]
@sethpollack wondering if there is anything we can learn from Tekton here

**Comment by sethpollack** [comment]
@bradrydzewski Yup, was wondering the same. Do you know anyone that we can reach out to?

**Comment by bradrydzewski** [comment]
@sethpollack I do not, however, there is some information [here](https://github.com/tektoncd/pipeline/tree/master/docs/developers#entrypoint-rewriting-and-step-ordering)

> Entrypoint is injected into the Task Container(s), wraps the Task step to manage the execution order of the containers. The entrypoint binary has the following arguments:
> 
> * wait_file - If specified, file to wait for
> * wait_file_content - If specified, wait until the file has non-zero size
> * post_file - If specified, file to write upon completion
> * entrypoint - The command to run in the image being wrapped
> As part of the PodSpec created by TaskRun the entrypoint for each Task step is changed to the entrypoint binary with the mentioned arguments and a volume with the binary and file(s) is mounted.

I think this is something we could easily replicate in Drone by wrapping the entrypoint of each pipeline step so that it executes the placeholder binary, which in turn waits for a file to updated in order to execute.  The question is, how do we get the placeholder binary inside every pipeline container?

**Comment by smira** [comment]
Entrypoint could be mounted as volume, and copied into that volume from init container in the same pod

**Comment by sethpollack** [comment]
We looked into using an entrypoint binary originally. The issue with that is that you would now need to specify the entrypoint for plugins rather than falling back to the entrypoint in the dockerfile.

**Comment by beniamin** [comment]
I am seeing similar issues. The step is stuck in the running state, because the image tag does not exists. The pod has 2 error events: `Error: ErrImagePull` and `Failed to pull image "my-private-image:v1": rcp error: code Unknown desc = Error response from the daemon: manifest for my-private-image:v1 not found: manifest unknown: manifest unkown`.

I don't know if the cause of my issue is the same as described above, but the behaviour is the same.

**Comment by willemveerman** [comment]
We have applied this change to tag [v1.0.0-beta.6](https://github.com/drone-runners/drone-runner-kube/releases/tag/v1.0.0-beta.6) and it has seemingly solved the problem of steps being stuck in a running state.

However, we have found that this change introduces a new issue of pipelines no longer respecting the timeout setting. If a step's duration causes the pipeline to exceed the timeout, the step is allowed to run to completion however no subsequent steps are run. By contrast, prior to this change, the whole pipeline was immediately terminated once the timeout was reached.

**Comment by bradrydzewski** [comment]
@marko-gacesa recently joined the team and he will be focusing on bringing the Kubernetes runner to GA.  He authored [this pull request](https://github.com/drone-runners/drone-runner-kube/pull/55) that redesigns how we watch for updates, including periodic polling as a fallback just in case the runner fails to receive an event from Kubernetes (for whatever reason).  This should make the runner much more resilient.  We would ask everyone in this thread to begin testing using the latest runner image.

This particular issue has been very difficult for us to triage because it is inconsistent; some people use the runner in production without issues and others periodically see pipelines or steps stuck in a running state.  Going forward when someone reports an issue we are going to ask for cluster details so that we can try to mirror your installation so that we can try to reproduce.  Furthermore, we may ask to jump on a Zoom call and debug the issue live.

Lets move further discussion to the following discourse thread:
https://discourse.drone.io/t/kubernetes-runner-stuck-in-running-state/9288

**Comment by marko-gacesa** [comment]
@smira, this pull request can no longer be merged as the container status monitoring code significantly diverged. I would like to close this. Can you please verify that the latest code solves issues with stuck builds. Stuck builds were happening because Kubernetes failed to run/terminate containers inside of a pod. In such situations the new implementation will abort builds with an error, instead of just getting stuck.

**Comment by smira** [comment]
> @smira, this pull request can no longer be merged as the container status monitoring code significantly diverged. I would like to close this. Can you please verify that the latest code solves issues with stuck builds. Stuck builds were happening because Kubernetes failed to run/terminate containers inside of a pod. In such situations the new implementation will abort builds with an error, instead of just getting stuck.

We plan on upgrading our CI cluster and we'll try new version, thank you
